source/common/grpc/typed_async_client.h

```diff
@@ -74,6 +83,42 @@ template <typename Response> class AsyncRequestCallbacks : public RawAsyncReques
   }
 };
 
+/**
+ * A versioned gRPC client.
+ */
+class VersionedClient {
+public:
+  virtual ~VersionedClient() = default;
+
+  /**
+   * @return std::string template of a fully-qualified service method name. For example:
+   * envoy.service.auth.{}.Authorization.Check.
+   */
+  virtual const std::string methodNameTemplate() const PURE;
+
+  /**
+   * @return std::string to override the namespace of a service for a specific version. When it
+   * returns non-EMPTY_STRING, methodNameTemplate() should return a template in the following
+   * format: envoy.service.{1}.{0}.SERVICE_NAME.METHOD_NAME. For example,
+   * envoy.service.{1}.{0}.Authorization.Check.
+   */
+  virtual const std::string serviceNamespace(envoy::config::core::v3::ApiVersion) const {
+    return EMPTY_STRING;
+  };
+
+  /**
+   * Given a version, return the method descriptor for a specific version.
+   *
+   * @param api_version target API version.
+   * @param use_alpha if this is an alpha version of an API client.
+   *
+   * @return Protobuf::MethodDescriptor of a method for a specific version.
+   */
+  const Protobuf::MethodDescriptor&
```

Comment:
It seems strange to expose out both the method name template and a way to achieve a concrete method descriptor from the same interface. It seems the former is an implementation of the latter. Or am I missing something?